Update the docs for JobManagedBy and JobPodReplacementPolicy related to pod termination (#46808)
kubernetes/website: content/en/docs/concepts/workloads/controllers/job.md

      reason: FailedIndexes
      status: "True"
      type: Failed
    - message: Job has failed indexes
      reason: FailedIndexes
      status: "True"

>       type: FailureTarget # Since v1.31

Author: Makes sense, especially as currently the 1.31 version is scattered in a couple of places. I think it makes sense to hint users that there is a change of default behavior, but one place is probably enough,

I have also flipped the order to the conditions to reflect the order of them being added in time.

```

{{< note >}}
Since v1.31, you will also observe in the status the `FailureTarget`
Job condition, with the same `reason` and `message` as for the the `Failed`
condition (see also [Job termination and cleanup](#job-termination-and-cleanup)).

> {{< /note >}}

Author: Makes sense, but I yet rephrased it a bit to make it more clear that it is `Failed` condition that takes reason and message from `FailureTarget`, not the other way round. 

Also, the case shown in this example is not describing failure due to pod failure policy, but backoff limit per index.

      type: Failed
```

> Note that, since v1.31, you will also observe in the status the `FailureTarget`

Reviewer: IIIUC, the Hugo shortcode would be better as you can see here: https://kubernetes.io/docs/contribute/style/style-guide/#shortcodes

Reviewer 2: `{{< skew currentVersion >}}`

That is, the Job termination mechanisms activated with `.spec.activeDeadlineSeconds`
and `.spec.backoffLimit` result in a permanent Job failure that requires manual intervention to resolve.

### Termination of Job pods

A terminated Job may be either succeeded or failed, and its status is reflected
by the Job conditions `Complete` or `Failed`, respectively. Prior to v1.31
these Job conditions are added when the Job termination process is triggered,
but some pods may still remain running at that point in time.

Since v1.31, when you enable either the `JobManagedBy` or
`JobPodReplacementPolicy` (enabled by default)
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/), the
Job controller awaits for termination of all pods before adding a condition
indicating that the Job is finished (either `Complete` or `Failed`).

Note that, the process of terminating all pods may take a substantial amount

> of time, depending the one Pod's `terminationGracePeriodSeconds` (see

Reviewer 3: ```suggestion
of time, depending on a Pod's `terminationGracePeriodSeconds` (see
```

Author: Done

[Pod termination](#docs/concepts/workloads/pods/pod-lifecycle/#pod-termination)),

> and thus adding the terminal the Job condition, even if the fate of the Job is

Reviewer 3: ```suggestion
and thus adding the terminal Job condition, even if the fate of the Job is
```

Author: Done

already determined.

> If you want to know, or the controller you write, the fate of the Job as soon

Reviewer 3: not sure if we need to specify the controller bit here

Author: Yeah, me neither, might be just noise. Dropped it.

as determined you can use, since v1.31, the `FailureTarget` and `SuccessCriteriaMet`
conditions, which cover all scenarios in which Job controller triggers
the Job termination process.

> 

Reviewer 3: crosslinking a discussion https://github.com/kubernetes/website/pull/46807#discussion_r1691543157

Author: +1. I have expanded on the `FailureTarget` and `SuccessCriteriaMet` conditions in the recent update. PTAL.

That is, the Job termination mechanisms activated with `.spec.activeDeadlineSeconds`
and `.spec.backoffLimit` result in a permanent Job failure that requires manual intervention to resolve.

### Terminal Job conditions

A Job has two possible terminal states, it ends up either succeeded, or failed,
and these states are reflected by the presence of the Job conditions `Complete`
or `Failed`, respectively.

The failure scenarios encompass:

> - the `.spec.backoffLimit`

Reviewer: ```suggestion
- the `.spec.backoffLimit` is reached
```

Maybe a full sentence would be better as similar to "the .spec.completions is reached".

Author: Yeah, but I think in case of backoffLimit "exceeded" is better. This is also what the reason on the Failed condition says.

Reviewer: That sounds more reasonable to me. Thanks!

- the `.spec.activeDeadlineSeconds` is exceeded
- the `.spec.backoffLimitPerIndex` is exceeded (see [Backoff limit per index](#backoff-limit-per-index))

> - the Pod matches the Job Pod Failure Policy rule with the `FailJob` action  (see more [Pod failure policy](#pod-failure-policy))

Reviewer 2: The spelling "Job Pod Failure Policy" doesn't conform to the website doc style.

Author: Updated to use the field name. PTAL.

The success scenarios encompass:
- the `.spec.completions` is reached

> - the criteria specified by the Job Success Policy are met (see more [Success policy](#success-policy))

Reviewer 2: "Job Success Policy" should be "Job success policy"?
"Success policy" should be "success policy"?

Reviewer: ```suggestion
- the criteria specified by the `.spec.successPolicy` are met (see more [Success policy](#success-policy))
```

Maybe, field name would be better. WDYT? @mimowo @tengqm

Reviewer 2: using generic terms or field names are both okay to me.

Author: Cool, I use the field name in this case here.

### Termination of Job pods

Prior to v1.31 the Job terminal conditions are added when the Job termination
process is triggered, and all Pod finalizers are removed, but some pods may
still remain running at that point in time.

Since v1.31, when you enable either the `JobManagedBy` or

> `JobPodReplacementPolicy` (enabled by default)

Reviewer 2: ```suggestion
the `JobPodReplacementPolicy` (enabled by default)
```

Author: Thanks, done

[feature gate](/docs/reference/command-line-tools-reference/feature-gates/), the
Job controller awaits for termination of all pods before adding a condition
indicating that the Job is finished (either `Complete` or `Failed`).

> Note that, the process of terminating all pods may take a substantial amount

Reviewer: Additionally, using the Hugo short code would be better, right?
But I'm not sure if we should use the Hugo shortcode in all places where we want to mention the Note.

Author: Actually, the paragraph seems pretty important for the section so tentatively I remove the "Note that", and don't add the Hugo either. PTAL.

Reviewer: SGTM

That is, the Job termination mechanisms activated with `.spec.activeDeadlineSeconds`
and `.spec.backoffLimit` result in a permanent Job failure that requires manual intervention to resolve.

### Terminal Job conditions

A Job has two possible terminal states, it ends up either succeeded, or failed,
and these states are reflected by the presence of the Job conditions `Complete`
or `Failed`, respectively.

The failure scenarios encompass:
- the `.spec.backoffLimit` is exceeded
- the `.spec.activeDeadlineSeconds` is exceeded

> - the `.spec.backoffLimitPerIndex` is exceeded (see [Backoff limit per index](#backoff-limit-per-index))

Author: Good point, added and rephrased the descriptions of existing ones. PTAL.

> - the Pod matches a rule in `.spec.podFailurePolicy` with the `FailJob` action (see more [Pod failure policy](#pod-failure-policy))

Author: > and mention here that `FailIndex` action can speed up the potential job failure when `backoffLimitPerIndex` is used?

I feel this level of detail should go to the dedicated sections. I prefer to keep here just one-liner descirptions, and more details in dedicated sections

> - the Pod matches a rule in `.spec.podFailurePolicy` with the `FailJob` action (see more [Pod failure policy](#pod-failure-policy))

Author: > Or just add that pod failure policy can further modify the failure scenarios?

Hm, you mean for example `action: Ignore` can influence handling of `spec.backoffLimit`? Yes, there are such cases, but I would prefer to keep the points here as concise one-liners. The users can find out more nuances when going to the specific sections. Capturing all nuances here will be essentially repeating the sections.

> - the Pod matches a rule in `.spec.podFailurePolicy` with the `FailJob` action (see more [Pod failure policy](#pod-failure-policy))

Author: I have rephrased the descriptions trying to keep the balance between being too concise and going into too much details. PTAL.

> - the Pod matches a rule in `.spec.podFailurePolicy` with the `FailJob` action (see more [Pod failure policy](#pod-failure-policy))

Reviewer 3: We do not have to go too much into details. What about something like https://github.com/kubernetes/website/pull/46808#discussion_r1693183116 ?

> - the Pod matches a rule in `.spec.podFailurePolicy` with the `FailJob` action (see more [Pod failure policy](#pod-failure-policy))

Author: Great. accepted the suggestion as it sounds good to me

The success scenarios encompass:

> - the `.spec.completions` is reached

Author: Yeah, I prefer the points here to be concise, but I agree the initial working was too cryptic. I have rephrased the points. PTAL.

> - the `.spec.completions` is reached

Reviewer 3: Thx

- the criteria specified by the `.spec.successPolicy` are met (see more [Success policy](#success-policy))

### Termination of Job pods

> Prior to {{< skew currentVersion >}} the Job terminal conditions are added when the Job termination

Reviewer 3: how does the `{{< skew currentVersion >}}` work? What happens when we release 1.32?

Author: @tengqm can you help with that?

Author: Ok, I tested this by looking at other places of its usage. For example this line "The kubelet in Kubernetes v1.30 collects spans from the garbage collection, pod synchronization routine as well as" in https://kubernetes.io/docs/concepts/cluster-administration/system-traces/ was added a year ago and it says "1.30", so there is no any mechanism to freeze the version.

I think our intention is to say "since v1.31", so probably it is better to hard-code it. Alternatively we can say "In Kubernetes {current version} ...". Which would you prefer @tengqm @atiratree ?

Author: Maybe the downside is that the documentation contains references to some very old versions like 1.21. 

OTOH, it is nice to know if some feature was introduced in 1.29 or 1.30, for example, if you are considering an upgrade.

Reviewer 3: I don't mind hard-coding it. Currently it would give a wrong implication in 1.32, that one needs to upgrade to latest to observe the behavior.

I suppose it would be good to remove the version mention in some number of releases as the version becomes less relevant. But that it quite hard to plan.

Author: I agree, the wording `Prior to {{< skew currentVersion >}}` or  `Since {{< skew currentVersion >}}` will be misleading. Tentatively I use hard-coding now.

process is triggered, and all Pod finalizers are removed, but some pods may

> still remain running at that point in time.

Reviewer 3: ```suggestion
still remain running/terminating at that point in time.
```

?

Author: sgtm, updated

That is, the Job termination mechanisms activated with `.spec.activeDeadlineSeconds`
and `.spec.backoffLimit` result in a permanent Job failure that requires manual intervention to resolve.

### Terminal Job conditions

A Job has two possible terminal states, it ends up either succeeded, or failed,
and these states are reflected by the presence of the Job conditions `Complete`

> or `Failed`, respectively.

Author: done

> The failure scenarios encompass:

Author: Actually, the lists are complete so I prefer to have a wording which reflects that. How about "The possible reasons for a Job failure / success". Tentatively updated this way.

- the number of pod failures exceeded the specified `.spec.backoffLimit`;
  see also [Pod backoff failure policy](#pod-backoff-failure-policy)
- the Job runtime exceeded the specified `.spec.activeDeadlineSeconds`
- the Job using `.spec.backoffLimitPerIndex` has some indexes failed;
  see also [Backoff limit per index](#backoff-limit-per-index)

> - the number of failed indexes in the Job using has exceeded the specified
>   `spec.maxFailedIndexes`; see also [Backoff limit per index](#backoff-limit-per-index)

Reviewer 3: ```suggestion
- the number of failed indexes in the Job has exceeded the specified
  `spec.maxFailedIndexes`; see also [Backoff limit per index](#backoff-limit-per-index)
```

> - a failed Pod matches a rule in `.spec.podFailurePolicy`
>    with the `FailJob` action; see also [Pod failure policy](#pod-failure-policy)

Reviewer 3: ```suggestion
- a failed Pod matches a rule in `.spec.podFailurePolicy`
   with the `FailJob` action; see also [Pod failure policy](#pod-failure-policy) for more details on how the policies can affect evaluation of failures
```

> The success scenarios encompass:

Author: Rephrased tentatively to "The possible reasons for a Job success" to better reflect this the complete list of scenarios.

- the number of succeeded pods reached the specified `.spec.completions`
- the criteria specified by the `.spec.successPolicy` are met;
  see also [Success policy](#success-policy)

### Termination of Job pods

Prior to v1.31 the Job terminal conditions are added when the Job termination
process is triggered, and all Pod finalizers are removed, but some pods may

> still remain running/terminating at that point in time.

Author: I think it would be useful to keep trace of the change in one place, as this is a substantial logic change, and even though the docs are versioned users may not have a clue something has changed. We did it in the past for numerous features, for example: https://kubernetes.io/docs/concepts/cluster-administration/node-shutdown/.

That being said, I make sure the version is mentioned in only one place (before review it was scattered across).

Since v1.31, when you enable either the
`JobManagedBy` or the `JobPodReplacementPolicy` (enabled by default)
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/), the
Job controller awaits for termination of all pods before adding a condition

> indicating that the Job is finished (either `Complete` or `Failed`).

Author: Rephrased

> The process of terminating all pods may take a substantial amount
> of time, depending on a Pod's `terminationGracePeriodSeconds` (see
> [Pod termination](#docs/concepts/workloads/pods/pod-lifecycle/#pod-termination)),
> and thus adding the terminal Job condition, even if the fate of the Job is
> already determined.
> 
> If you want to know the fate of the Job as soon as determined you can use,
> since v1.31, the `FailureTarget` and `SuccessCriteriaMet` conditions, which
> cover all scenarios in which Job controller triggers the Job termination process
> (see [Terminal Job conditions](#terminal-job-conditions)).

Reviewer 4: ```suggestion
Factors like `terminationGracePeriodSeconds` might increase the amount of time
from the moment that the Job controller adds the `FailureTarget` condition or the
`SuccessCriteriaMet` condition to the moment that all of the Job Pods terminate
and the Job controller adds a [terminal condition](#terminal-job-conditions)
(`Failed` or `Complete`).

You can use the `FailureTarget` or the `SuccessCriteriaMet` condition to evaluate
whether the Job has failed or succeeded without having to wait for the controller
to add a terminal condition.
```

Reviewer 4: I think that this is the most important bit of info. You can also add an example or two to the last paragraph for WHY you'd need this, like 

"For example, you can use this information to quickly decide whether to create a replacement Job, or to make scheduling decisions based on node availability."

Author: Rephrased, also added the WHY in a note.

### Terminal Job conditions

A Job has two possible terminal states, each of which has a corresponding Job
condition:
* Succeeded:  Job condition `Complete`
* Failed: Job condition `Failed`.

The possible reasons for a Job failure:
- the number of pod failures exceeded the specified `.spec.backoffLimit`;
  see also [Pod backoff failure policy](#pod-backoff-failure-policy)
- the Job runtime exceeded the specified `.spec.activeDeadlineSeconds`
- the Job using `.spec.backoffLimitPerIndex` has some indexes failed;
  see also [Backoff limit per index](#backoff-limit-per-index)
- the number of failed indexes in the Job has exceeded the specified
  `spec.maxFailedIndexes`; see also [Backoff limit per index](#backoff-limit-per-index)
- a failed Pod matches a rule in `.spec.podFailurePolicy`
   with the `FailJob` action; see also [Pod failure policy](#pod-failure-policy) for more details on how the policy rules can affect evaluation of failures

The possible reasons for a Job success:
- the number of succeeded pods reached the specified `.spec.completions`
- the criteria specified by the `.spec.successPolicy` are met;
  see also [Success policy](#success-policy)

> Since v1.31 the Job controller delays the addition of the terminal conditions

Reviewer 5: ```suggestion
In Kubernetes v1.31 and later the Job controller delays the addition of the terminal conditions
```
?

> Since v1.31 the Job controller delays the addition of the terminal conditions

Reviewer 3: ```suggestion
Since v1.31 the Job controller delays the addition of the terminal conditions `Complete` or `Failed`
```

I think it wouldn't hurt to reiterate the conditions.

Author: sure, done

until all pods are terminated.

{{< note >}}

> In the previous releases Job terminal conditions are added when the Job

Reviewer 5: ```suggestion
In Kubernetes v1.30 and earlier, Job terminal conditions were added when the Job
```
Use past tense ideally.

and `.spec.backoffLimit` result in a permanent Job failure that requires manual intervention to resolve.

### Terminal Job conditions

A Job has two possible terminal states, each of which has a corresponding Job
condition:
* Succeeded:  Job condition `Complete`
* Failed: Job condition `Failed`.

The possible reasons for a Job failure:
- The number of Pod failures exceeded the specified `.spec.backoffLimit` in the Job
  specification. For details, see [Pod backoff failure policy](#pod-backoff-failure-policy).
- The Job runtime exceeded the specified `.spec.activeDeadlineSeconds`
- An indexed Job that used `.spec.backoffLimitPerIndex` has failed indexes.
  For details, see [Backoff limit per index](#backoff-limit-per-index).
- The number of failed indexes in the Job exceeded the specified
  `spec.maxFailedIndexes`. For details, see [Backoff limit per index](#backoff-limit-per-index)
- A failed Pod matches a rule in `.spec.podFailurePolicy` that has the `FailJob`
   action. For details about how Pod failure policy rules might affect failure
   evaluation, see [Pod failure policy](#pod-failure-policy).

> The possible reasons for a Job success:

Author: done

- The number of succeeded Pods reached the specified `.spec.completions`
- The criteria specified in `.spec.successPolicy` are met. For details, see
  [Success policy](#success-policy).

In Kubernetes v1.31 and later the Job controller delays the addition of the

> terminal conditions,`Failed` or `Succeeded`, until all pods are terminated.

Reviewer 4: ```suggestion
terminal conditions,`Failed` or `Complete`, until all of the Job Pods are terminated.
```

> {{< note >}}

Reviewer 4: I think this doesn't belong in a note, because the previous sentence is also related to the content inside the note.

Reviewer 4: Or bring the preceding sentence about 1.31 and later into the note block

Author: Ok, I chose to remove the note entirely

> In Kubernetes v1.30 and earlier, Job terminal conditions were added when the Job
> termination process is triggered, and all Pod finalizers are removed, but some
> pods may still remain running/terminating at that point in time.
> 
> The change of the behavior is activated by enablement of the `JobManagedBy` or
> `JobPodReplacementPolicy` (enabled by default)
> [feature gates](/docs/reference/command-line-tools-reference/feature-gates/).

Reviewer 4: ```suggestion
In Kubernetes v1.30 and earlier, the Job controller added the `Complete` or the
`Failed` Job terminal conditions as soon as the Job termination process was
triggered and all Pod finalizers were removed. However, some Pods would still
be running or terminating at the moment that the terminal condition was added.

In Kubernetes v1.31 and later, the controller only adds the Job terminal conditions
_after_ all of the Pods are terminated. You can enable this behavior by using the `JobManagedBy` or the `JobPodReplacementPolicy` (enabled by default)
[feature gates](/docs/reference/command-line-tools-reference/feature-gates/).
```

{{< /note >}}

### Termination of Job pods

The Job controller adds the `FailureTarget` condition or the `SuccessCriteriaMet`
condition to the Job to trigger Pod termination after a Job meets either the
success or failure criteria.

Factors like `terminationGracePeriodSeconds` might increase the amount of time
from the moment that the Job controller adds the `FailureTarget` condition or the
`SuccessCriteriaMet` condition to the moment that all of the Job Pods terminate
and the Job controller adds a [terminal condition](#terminal-job-conditions)
(`Failed` or `Complete`).

You can use the `FailureTarget` or the `SuccessCriteriaMet` condition to evaluate
whether the Job has failed or succeeded without having to wait for the controller
to add a terminal condition.

> {{< note >}}

Reviewer 4: No need a note callout; the example flows nicely with the rest of the section

Author: Ok, removed.

> For example, you can use the `FailureTarget` condition to quickly decide whether
> to create a replacement Job, but it could result in Pods from the failing and
> replacement Jobs running at the same time for a while. Thus, if your cluster
> capacity is limited, you may prefer to wait for the `Failed` condition before
> creating the replacement Job.

Reviewer 4: ```suggestion
For example, you might want to decide when to create a replacement Job
that replaces a failed Job. If you replace the failed Job when the `FailureTarget`
condition appears, your replacement Job runs sooner, but could result in Pods
from the failed and the replacement Job running at the same time, using
extra compute resources. 

Alternatively, if your cluster has limited resource capacity, you could choose to
wait until the `Failed` condition appears on the Job, which would delay your
replacement Job but would ensure that you conserve resources by waiting
until all of the failed Pods are removed.
```

  terminating: 3 # three Pods are terminating and have not yet reached the Failed phase
```

{{< note >}}

> Since v1.31 the `JobPodReplacementPolicy` feature gate guards the fix for the issue:

Reviewer 5: We prefer to link to the list of feature gates:
```suggestion
Since v1.31 the `JobPodReplacementPolicy` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) guards the fix for the issue:
```

Author: For sure, it was just a quick note. I m going to work on a proper documentation update this week.

Author: Updated. PTAL

  terminating: 3 # three Pods are terminating and have not yet reached the Failed phase
```

{{< note >}}

> Since {{< skew currentVersion >}}, when you enable the `JobPodReplacementPolicy`

Reviewer 3: same question about the version in other places as well

Author: Yeah, let's track the discussion there: https://github.com/kubernetes/website/pull/46808#discussion_r1692964346, and I will update all places if needed.

  terminating: 3 # three Pods are terminating and have not yet reached the Failed phase
```

{{< note >}}
Since v1.31, when you enable the `JobPodReplacementPolicy`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
(enabled by default), the Job controller awaits for termination of all pods
before marking a Job as terminal (see [Termination of Job Pods](#termination-of-job-pods)).

> {{< /note >}}

Reviewer 4: Not sure how this relates to the replacement of failed Pods. I would remove this, since it's a bit confusing here

Author: I agree, it can be confusing. The only purpose of this was to reference that the logic change is controlled by the feature gate. Now I propose to mention this as a note in one place only.

indicated by the field value.
{{< /warning >}}

> {{< note >}}

Reviewer 4: Note not needed in this section, as it's not really related to the section's topic

Author: I agree, it can be confusing. The only purpose of this was to reference that the logic change is controlled by the feature gate. Now I propose to mention this as a note in one place only.
